tianshou/utils/net/common.py

```diff
@@ -159,3 +164,55 @@ def forward(
         # please ensure the first dim is batch size: [bsz, len, ...]
         return s, {"h": h.transpose(0, 1).detach(),
                    "c": c.transpose(0, 1).detach()}
+
+
+class CategoricalNet(Net):
+    """Simple MLP backbone.
+
+    For advanced usage (how to customize the network), please refer to
+    :ref:`build_the_network`.
+
+    .. seealso::
+
+        Please refer to :class:`~tianshou.utils.net.common.Net` for
+         more detailed explanation.
+    """
+
+    def __init__(
+        self,
+        layer_num: int,
+        state_shape: tuple,
+        action_shape: Optional[Union[tuple, int]] = 0,
+        device: Union[str, int, torch.device] = "cpu",
+        concat: bool = False,
+        hidden_layer_size: int = 128,
+        dueling: Optional[Tuple[int, int]] = None,
+        norm_layer: Optional[Callable[[int], nn.modules.Module]] = None,
+        num_atoms: int = 51,
+    ) -> None:
+        super().__init__(layer_num, state_shape, action_shape,
+                         device, True, concat, hidden_layer_size,
+                         dueling, norm_layer, num_atoms)
+        self.action_shape = action_shape
+        self.num_atoms = num_atoms
+
+    def forward(
+        self,
+        s: Union[np.ndarray, torch.Tensor],
+        state: Optional[Dict[str, torch.Tensor]] = None,
+        info: Dict[str, Any] = {},
+    ) -> Tuple[torch.Tensor, Dict[str, torch.Tensor]]:
+        """Mapping: s -> flatten -> logits."""
+        s = to_torch(s, device=self.device, dtype=torch.float32)
+        s = s.reshape(s.size(0), -1)
+        logits = self.model(s)
+        if self.dueling is not None:  # Dueling DQN
+            q, v = self.Q(logits), self.V(logits)
+            v = v.view(-1, 1, self.num_atoms),
+            q = q.view(-1, np.prod(self.action_shape), self.num_atoms)
+            logits = q - q.mean(dim=1, keepdim=True) + v
+        else:
+            logits = logits.view(
+                -1, np.prod(self.action_shape), self.num_atoms)
+        logits = torch.softmax(logits, dim=-1)
```

Comment:
Can these lines be simplified? 
```python
logits, state = super().forward(s, state, info)
logits = logits.view(-1, np.prod(self.action_shape), self.num_atoms)
if self.softmax:
    logits = torch.softmax(logits, dim=-1)
return logits, state
```

Reply:
I don't think so. The mean of q in c51 is of size (batch_size, num_atoms). However, the mean of q in dqn is of size (batch_size) .

Reply:
Got it

Reply:
Maybe we can change the way of calculating the mean of q in dqn. But it may need extra judgment that whether num_atoms is 1.